### PR TITLE
[quant] creating quint4x2 dtype for quantized tensors

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45594 [quant] PerChannelFloatQParams support for quint4x2 dtype
* **#44678 [quant] creating quint4x2 dtype for quantized tensors**

Summary:
This is a prototype PR that introduces 4 bit qtensors. The new dtype added for this is c10::quint4x2
The underlying storage for this is still uint8_t, so we pack 2 4-bit values in a byte while quantizing it.

This change uses most of the existing scaffolding for qtensor storage. We allocate storage
based on the dtype before creating a new qtensor.

It also adds a dispatch mechanism for this dtype so we can use this to get the bitwidth, qmin and qmax info
while quantizing and packing the qtensor (when we add 2-bit qtensor)

Kernels that use this dtype should be aware of the packing format.

Test Plan:
Locally tested
```
x = torch.ones((100, 100), dtype=torch.float)
qx_8bit = torch.quantize_per_tensor(x, scale=1.0, zero_point=2, dtype=torch.quint8)
qx = torch.quantize_per_tensor(x, scale=1.0, zero_point=2, dtype=torch.quint4x2)

torch.save(x, "temp.p")
print('Size float (B):', os.path.getsize("temp.p"))
os.remove('temp.p')

torch.save(qx_8bit, "temp.p")
print('Size quantized 8bit(B):', os.path.getsize("temp.p"))
os.remove('temp.p')

torch.save(qx, "temp.p")
print('Size quantized 4bit(B):', os.path.getsize("temp.p"))
os.remove('temp.p')
```

Size float (B): 40760
Size quantized 8bit(B): 10808
Size quantized 4bit(B): 5816

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23993134](https://our.internmc.facebook.com/intern/diff/D23993134)

### CODE REVIEW

File: aten/src/ATen/Dispatch.h [hunk 1]
```diff
@@ -346,6 +361,25 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                       \
   }()
 
+#define AT_DISPATCH_QINT_AND_SUB_BYTE_TYPES(TYPE, NAME, ...)                                   \
```
Comment (jerryzh168) [inline aten/src/ATen/Dispatch.h:364]: can we merge this with AT_DISPATCH_QINT_TYPES?

File: aten/src/ATen/native/quantized/affine_quantizer.cpp [hunk 2]
```diff
@@ -103,13 +106,21 @@ Tensor quantize_tensor_per_tensor_affine(
...
+  AT_DISPATCH_QINT_AND_SUB_BYTE_TYPES(qtensor.scalar_type(), fn_name, [&]() {
     checkQuantizedTensor<scalar_t>(fn_name, qtensor);
     checkZeroPoint<underlying_t>(fn_name, zero_point);
   });
 
-  quantize_tensor_per_tensor_affine_stub(
+  // Temporary solution to pack the tensor if dtype is torch.quint4x2
+  // Can move this into the fbgemm::Quantize op.
+  if (qtensor.scalar_type() == at::ScalarType::QUInt4x2) {
+    quantize_tensor_per_tensor_affine_sub_byte_stub(
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/affine_quantizer.cpp:117]: looks like this has the same signature as quantize_tensor_per_tensor_affine_stub can we merge them?

File: aten/src/ATen/Dispatch.h (hunk 1 above)
Comment (supriyar) [reply aten/src/ATen/Dispatch.h:364]: Not currently. The AT_DISPATCH_QINT_TYPES is used for other kernels as well and if we try to dispatch over quint4x2, there will be errors in vec256 not being able to understand the scalar type

File: aten/src/ATen/native/quantized/affine_quantizer.cpp (hunk 2 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/affine_quantizer.cpp:117]: The current quantize_per_tensor_affine_stub calls fbgemm::Quantize function, once we are able to support 4bit in fbgemm quantize we can move it there and delete the new stub.

File: test/quantization/test_quantized_tensor.py [hunk 3]
```diff
@@ -103,6 +103,36 @@ def test_qtensor(self):
                                  "quantization_scheme=torch.per_tensor_affine, " +
                                  "scale=1.0, zero_point=2)")
 
+    def test_qtensor_sub_byte(self):
+        num_elements = 10
+        scale = 1.0
+        zero_point = 2
```
Comment (raghuramank100) [inline test/quantization/test_quantized_tensor.py:109]: For 4 bit and 2 bit tensors, we should plan on having floating point scale and zero-point, as this format is primarily for lower precision storage.

File: test/quantization/test_quantized_tensor.py [hunk 4]
```diff
@@ -216,15 +246,10 @@ def test_qtensor_dtypes(self):
...
-        rqr = qr.dequantize()
-        self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / scale))
-        qr = torch.quantize_per_tensor(r, scale, zero_point, torch.quint8)
-        rqr = qr.dequantize()
-        self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / scale))
-        qr = torch.quantize_per_tensor(r, scale, zero_point, torch.qint32)
-        rqr = qr.dequantize()
-        self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / scale))
+        for dtype in [torch.qint8, torch.quint8, torch.qint32, torch.quint4x2]:
+            qr = torch.quantize_per_tensor(r, scale, zero_point, dtype)
```
Comment (raghuramank100) [inline test/quantization/test_quantized_tensor.py:250]: Can we use the same API with floating point scale and zero-point for qint4x2?

File: test/quantization/test_quantized_tensor.py [hunk 5]
```diff
@@ -216,15 +246,10 @@ def test_qtensor_dtypes(self):
...
-        qr = torch.quantize_per_tensor(r, scale, zero_point, torch.qint8)
-        rqr = qr.dequantize()
-        self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / scale))
-        qr = torch.quantize_per_tensor(r, scale, zero_point, torch.quint8)
-        rqr = qr.dequantize()
-        self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / scale))
-        qr = torch.quantize_per_tensor(r, scale, zero_point, torch.qint32)
-        rqr = qr.dequantize()
-        self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / scale))
+        for dtype in [torch.qint8, torch.quint8, torch.qint32, torch.quint4x2]:
```
Comment (raghuramank100) [inline test/quantization/test_quantized_tensor.py:249]: Also, is there a plan to add quantize_per_channel in a later PR? For lower precisions, per row/per-channel quantization will be more useful.

File: test/quantization/test_quantized_tensor.py (hunk 4 above)
Comment (supriyar) [reply test/quantization/test_quantized_tensor.py:250]: I can add this support in a follow-up PR. Currently floating point scale and zero point support was added for per-channel quantization, so I can extend that to support this dtype as well.

File: test/quantization/test_quantized_tensor.py (hunk 5 above)
Comment (supriyar) [reply test/quantization/test_quantized_tensor.py:249]: Yes, please see my comment above. I can follow up with support for per-channel with floating point zero-point.

**Review by raghuramank100 — APPROVED** [review]

**Review by jerryzh168 — APPROVED** [review]

**Pull request closed without merging**